external_executors/dnanexus_wes/dx-wes-lambda/dx_wes.py

```diff
@@ -0,0 +1,510 @@
+        new DXFile.
+        """
+
+        # Google urls can be converted to https
+        if url.startswith("gs://"):
+            client = google.cloud.storage.client.Client.create_anonymous_client()
+
+            path_parts = url.replace("gs://", "").split("/")
+            bucket_name = path_parts[0]
+            bucket = client.bucket(bucket_name)
+            blob_name = "/".join(path_parts[1:])
+            blob = bucket.blob(blob_name)
+            public_url = blob.public_url
+        elif url.startswith("https://") or url.startswith("http://"):
+            public_url = url
+        else:
+            return url
+
+        # Create a new DXFile object and leave it open for writing
+        dx_file = dxpy.new_dxfile(
+            name=os.path.basename(url),
+            mode="w",
+            folder='/' + wes_workflow_id,
+            parents=True,
+            project=dx_project_id)
+
+        # Stream the remote file into the DXFile
+        response = requests.get(public_url, stream=True)
+        for chunk in response.iter_content(chunk_size=1<<24):
+            dx_file.write(chunk)
+
+        # Close and return the ID prefixed with the new scheme, dx
+        dx_file.close()
+
+        return 'dx://' + dx_file.id
+
+    def localize_inputs(obj, match_func, localize_func):
+        """Recursively walk through the input dict, localizing inputs as
+        necessary.
+        """
+
+        if isinstance(obj, (list, tuple)):
+            return [localize_inputs(e, match_func, localize_func) for e in obj]
+        if isinstance(obj, dict):
+            return {k: localize_inputs(v, match_func, localize_func) for k, v in obj.items()}
+        if isinstance(obj, basestring):
+            if match_func(obj):
+                return localize_func(obj)
+        return obj
+
+    localized_inputs_dict = localize_inputs(inputs_dict, match_url, localize_url)
+
+    return localized_inputs_dict
+
+
+def dx_to_wes_state(dx_state):
+    """Convert the state returned by DNAnexus to something from the list of states
+    defined by WES.
+    """
+
+    if dx_state == "in_progress":
+        return "RUNNING"
+    elif dx_state in ("partially_failed", "failed"):
+        return "EXECUTOR_ERROR"
+    elif dx_state in "done":
+        return "COMPLETE"
+
+    return "UNKNOWN"
+
+
+##############################################################################
+## Lambda functions
+##############################################################################
+
+def run_dxwdl(event, context):
+    """Run dxWDL to create a DNAnexus workflow.
+
+    This function is invoked by the function that handles POST to /workflows.
+    We'd like to expose that function via API Gateway, but API Gateway has a
+    29 second limit on a response, and dxWDL and input localization may not
+    complete in that time. But, it's pretty likely to complete in the five
+    minute lambda time limit. So, we delegate all the actual work here and
```

Comment:
This is just localizing the files necessary to initiate a run, right? We're not talking about data files yet, as that's handled by dxWDL downstream?

Reply:
It's doing both. It first copies files to dnanexus and gets dnanexus ids for them. Then it gives those ids to dxWDL so it can create a workflow with the inputs already fixed. The workflow that gets run on dnanexus will only know about the ids of the localized files

Reply:
Ah wow, so you need to grab all the inputs to the first task? What happens if you ask for a huge bam file (e.g. a full-scale `10xv2` run)? 

... I could see the download taking too long, and things timing out.

Reply:
Yeah, this is a very bad thing. I guess we could just start a real instance and run everything on that, but then we're so far from the vision of a lightweight WES shim.

Reply:
Okay after some discussion with Andrey, the best way forward appeared to be shifting as much as possible into dnanexus itself rather than within the translation layer. So this PR has been updated (quite a bit) to move all the file transfers and dxWDL execution into dnanexus jobs.